# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmCodegenTypes.h
# Commit: d6506b5ee07b
# Full Hash: d6506b5ee07b5c65f889ce7c2214b0172f205898
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-16 09:43:03
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r 80a164c1054e -r d6506b5ee07b js/src/wasm/WasmCodegenTypes.h
--- a/js/src/wasm/WasmCodegenTypes.h	Wed Sep 15 11:02:57 2021 +0000
+++ b/js/src/wasm/WasmCodegenTypes.h	Wed Sep 15 11:10:15 2021 +0000
@@ -245,7 +245,9 @@
     TrapExit,          // calls C++ to report and jumps to throw stub
     DebugTrap,         // calls C++ to handle debug event
     FarJumpIsland,     // inserted to connect otherwise out-of-range insns
-    Throw              // special stack-unwinding stub jumped to by other stubs
+    Throw,             // special stack-unwinding stub jumped to by other stubs
+    IndirectStub,  // a stub that take care of switching instance specific state
+                   // at cross-instance boundaries
   };
 
  private:
@@ -327,8 +329,9 @@
   bool isJitEntry() const { return kind() == JitEntry; }
   bool isInterpEntry() const { return kind() == InterpEntry; }
   bool isEntry() const { return isInterpEntry() || isJitEntry(); }
+  bool isIndirectStub() const { return kind() == IndirectStub; }
   bool hasFuncIndex() const {
-    return isFunction() || isImportExit() || isEntry();
+    return isFunction() || isImportExit() || isEntry() || isIndirectStub();
   }
   uint32_t funcIndex() const {
     MOZ_ASSERT(hasFuncIndex());
@@ -410,7 +413,8 @@
 
   enum Kind {
     Func,        // pc-relative call to a specific function
-    Dynamic,     // dynamic callee called via register
+    Import,      // wasm import call
+    Indirect,    // wasm indirect call
     Symbolic,    // call to a single symbolic callee
     EnterFrame,  // call to a enter frame handler
     LeaveFrame,  // call to a leave frame handler
@@ -427,7 +431,8 @@
   }
   uint32_t lineOrBytecode() const { return lineOrBytecode_; }
   Kind kind() const { return Kind(kind_); }
-  bool mightBeCrossInstance() const { return kind() == CallSiteDesc::Dynamic; }
+  bool isImportCall() const { return kind() == CallSiteDesc::Import; }
+  bool isIndirectCall() const { return kind() == CallSiteDesc::Indirect; }
 };
 
 class CallSite : public CallSiteDesc {